### PR TITLE
Add Prepayments

### LINKED ISSUE #808: (feat) Handling automatically the CAUTION when invoicing a patient
This issue is about to handle CAUTION of a patient automatically when invoicing the patient.

We must have a mean to signal to the user of BHIMA who makes invoice that the patient has a CAUTION and if we have to use this CAUTION for an invoice.

### Caution Steps: 

1. Client pay a caution via the `cash payment module`

    - The client will be `credited`
    - The cashbox will be `debited`
    - A caution receipt is generated as proof 

2. Bill the patient : 

    - Display the remaining caution amount, in reality the balance of the client which is `< 0`
    - In this step, the caution displayed is for `just notify` the user that the enterprise must money to the client
    - The caution amount doesn't affect the invoice, invoices are independent to cautions

3. Patient Financial Activities : 

    - Notify also for a remaining caution amount of the client

### LINKED ISSUE #1324: (feature request) Caution refactor and tests
There are a number of issues tracking the current implementation of cautions in 2.x. This must be corrected before releasing a solid financial base - many institutions rely on maintaining a line of credit with patients. 

Relevant discussions: 
https://github.com/IMA-WorldHealth/bhima-2.X/issues/808
https://github.com/Vanga-Hospital/bhima-2.X/issues/121
https://github.com/Vanga-Hospital/bhima-2.X/issues/34

### LINKED ISSUE #2687: Design: Caution "Links" in 2.x
This issue describes as design for balancing caution payments in 2.x

Previous issues on cautions: https://github.com/IMA-WorldHealth/bhima-2.X/issues/808, https://github.com/IMA-WorldHealth/bhima-2.X/issues/1324.

A caution payment is any payment that happens _before_ goods or services are rendered.  These are often required as a down-payment or collateral to ensure the hospital can recuperate some of its investment before engaging in an expensive operation.

**The Problem of Cautions in 2.x**
Cautions are tricky in 2.x because of how we track balances in invoices and patients.  A patient's balance is computed by looking at the `entity_uuid` column.  This does not take into account what kinds of transactions hit the debtor's account - it only knows the debtor account was it.

However, 2.x also tracks the amount due on an invoice by patient.  This is different from the patient balance owed to the enterprise, since it groups by `reference_uuid`.  All "unbalanced" invoices are shown for payment at the Cash Payment windows, even if the debtor's overall balance is different from the sum of their invoices if they paid a caution.

**Linking Cautions**
The problem is clear - cautions are never linked to the invoices, making them unbalanced, from BHIMA's perspective, regardless of the debtor's situation.

I propose that we implement a caution "link" that associates one (or more) caution(s) with an invoice at the time of invoice creation.  This "caution link" will be a voucher created _after_ the invoice is created that associates the patient's invoice with their previous payment.  It will look like this:

| **Record** | **Account** | **Debit** | **Credit** | **Entity** | **Description** | **Reference** |
| --------- | ----- | -----:| ----:| ---- | ---- | --- |
| **CP.TPA.1** | 41XXXX | 0 | $50.00 | PA.TPA.2 | A caution payment of $50.00 for PA.TPA.2 | |
| **CP.TPA.1** | 57XXXX | $50.00 | 0 |   | A caution payment of $50.00 for PA.TPA.2 | |
| **IV.TPA.1** | 41XXXX | $15.00 | 0 | PA.TPA.2 | An invoice to PA.TPA.2 for quinine and penicillin | |
| **IV.TPA.1** | 75XXXX | 0 | $7.00 |   | An invoice to PA.TPA.2 for quinine and penicillin | |
| **IV.TPA.1** | 75XXXX | 0 | $8.00 |   | An invoice to PA.TPA.2 for quinine and penicillin | |
| **VO.TPA.1** | 41XXXX | $15.00 | 0 | PA.TPA.2 | Caution link for IV.TPA.1 to CP.TPA.1 | **CP.TPA.1** |
| **VO.TPA.1** | 41XXXX | 0 | $15.00 | PA.TPA.2 | Caution link for IV.TPA.1 to CP.TPA.1 | **IV.TPA.1** |

**Note:** According to BHIMA's current rules, this transaction is invalid.  A transaction should not involved only a single account.  We would need to relax this constraint - either only for these "caution links" or in general.

The end result of the above this the following:
 1. Balance on Debtor's Account from summing all the rows with `entity_uuid` equal to PA.TPA.2:

| **Record** | **Account** | **Debit** | **Credit** | **Entity** | **Description** | **Reference** |
| --------- | ----- | -----:| ----:| ---- | ---- | --- |
| **CP.TPA.1** | 41XXXX | 0 | $50.00 | PA.TPA.2 | A caution payment of $50.00 for PA.TPA.2 | |
| **IV.TPA.1** | 41XXXX | $15.00 | 0 | PA.TPA.2 | An invoice to PA.TPA.2 for quinine and penicillin | |
| **VO.TPA.1** | 41XXXX | $15.00 | 0 | PA.TPA.2 | Caution link for IV.TPA.1 to CP.TPA.1 | **CP.TPA.1** |
| **VO.TPA.1** | 41XXXX | 0 | $15.00 | PA.TPA.2 | Caution link for IV.TPA.1 to CP.TPA.1 | **IV.TPA.1** |
|    |      | **$30.00** | **$65.00** | **$35.00** |  |  |

Therefore, the debtor has $35.00 as their total balance.  Notice how the "caution link" inflates the debits/credits slightly.

 2. Balance owed by `PA.TPA.2` on `IV.TPA.1`:

| **Record** | **Account** | **Debit** | **Credit** | **Entity** | **Description** | **Reference** |
| --------- | ----- | -----:| ----:| ---- | ---- | --- |
| **IV.TPA.1** | 41XXXX | $15.00 | 0 | PA.TPA.2 | An invoice to PA.TPA.2 for quinine and penicillin | |
| **VO.TPA.1** | 41XXXX | 0 | $15.00 | PA.TPA.2 | Caution link for IV.TPA.1 to CP.TPA.1 | **IV.TPA.1** |
|    |  |  **$15.00** | **$15.00** | **$0.00**  |   |  |

Therefore, the invoice is balanced.

### LINKED ISSUE #2695: Proposal: make Prepayments/Caution payments an enterprise setting
We added "enterprise settings" in https://github.com/IMA-WorldHealth/bhima-2.X/pull/2607 to toggle features on and off behaviors and features in the BHIMA application as a whole.  So far, we've only used this for allowing price changes in the Patient Invoice module.

I propose that we add an additional setting to disable caution payments.  One of our install points (HEV) does not use cautions at all.  We could save some logic, rendering, and avoid mistakes by disabling them completely if they are not used.

The basic idea would be to put a flag "enable_prepayments" in the `enterprise_setting` table.  This flag would control the following things:
 1. The "caution" option would not be available on the Cash page.  The page would default to "invoice payment".
 2. The alerts for creditor balances on the Patient Invoice page would never show, even if there was a creditor balance.
 3. The server would not check a creditor balance and skip logic described in #2687.

### PR DESCRIPTION
**Pull request by jniles**
This PR adds the feature for prepayments/caution payments to the main 2.x codebase.  It can be toggle on and off via a settings flag in the enterprise settings page.

The full description is provided in https://github.com/IMA-WorldHealth/bhima-2.X/issues/2687.

 1. Closes #808 
 2. Closes #1324
 3. Closes #2687
 4. Closes #2695

### CODE REVIEW

**Comment by jniles** [comment]
@mbayopanda, can you give this one final test round to see if it works to your satisfaction?

File: client/src/modules/patients/patients.service.js [hunk 1]
```diff
@@ -75,8 +59,8 @@ function PatientService(
    * @param {String} uuid The patient's UUID
    */
   function balance(uuid) {
-    const path = 'patients/:uuid/finance/balance';
-    return service.$http.get(path.replace(':uuid', uuid))
+    const path = `/patients/${uuid}/finance/balance`;
```
Comment (mbayopanda) [inline client/src/modules/patients/patients.service.js:62]: 👍 nice

**Review by mbayopanda — APPROVED** [review]
LGTM 👍 

bors r+

**Pull request merged**